src/typechecking.rs

```diff
@@ -45,10 +45,13 @@ impl Default for TypeInfo {
         res.presort_names.extend(MapSort::presort_names());
         res.presort_names.extend(SetSort::presort_names());
         res.presort_names.extend(VecSort::presort_names());
+        res.presort_names.extend(FunctionSort::presort_names());
```

Comment:
Should we make typechecking work for this PR? What happens if you try to run an ill-typed program- do you get ill-typed eclasses?

Reply:
I added a few type checking failure examples to show that it will make sure the input types are correct when calling a function and also that the return type is properly typed as well. I don't believe there is a way to get ill typed e-classes with this.

Reply:
Thanks! Nicely done